zjit/src/hir.rs

```diff
@@ -4169,6 +4170,9 @@ impl Function {
             | Insn::DefinedIvar { self_val: val, .. } => {
                 self.assert_subtype(insn_id, val, types::BasicObject)
             }
+            Insn::GuardNotFrozen { recv, .. } => {
+                self.assert_subtype(insn_id, recv, types::HeapBasicObject)
```

Comment:
Neat way to capture the precondition :)
Didn't know about `HeapBasicObject`, we could use that as the return type of `GuardNotFrozen` (even before this PR) but now that's automatic thanks to `Insn::GuardNotFrozen { recv, .. } => self.type_of(*recv),`

Reply:
That's true, didn't even think about that refinement beforehand!